ambry-account/src/main/java/com/github/ambry/account/JsonAccountService.java

```diff
@@ -0,0 +1,235 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.account;
+
+import com.github.ambry.config.JsonAccountConfig;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileTime;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Random;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Account service which uses a local JSON file as source for accounts.
+ *
+ * This account implementation is primarily useful for small deployments and development. The service operates
+ * completely in memory. Meaning that the entire JSON file is parsed and loaded in to memory.
+ */
+final class JsonAccountService extends AbstractAccountService {
+
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  /** Location of the account file which this service uses as it's source of accounts.*/
+  private final Path accountFile;
+
+  private final ScheduledExecutorService scheduler;
+
+  private final AccountServiceMetrics accountServiceMetrics;
+
+  private final JsonAccountConfig accountConfig;
+
+  /**
+   * Constructs an {@code JsonAccountService}. {@code #init} method MUST be called after construction.
+   *
+   * @param accountFile The path to the JSON file containing all the accounts.
+   * @param accountServiceMetrics Metrics instance to log metrics to.
+   * @param scheduler Scheduler used to periodically check the local account JSON for changes. When {@code null}
+   *                  accounts file will only be loaded once during startup.
+   */
+  JsonAccountService(Path accountFile, AccountServiceMetrics accountServiceMetrics, ScheduledExecutorService scheduler,
+      JsonAccountConfig accountConfig) {
+    super(accountServiceMetrics);
+
+    this.accountFile = Objects.requireNonNull(accountFile, "helixStore cannot be null");
+    this.accountServiceMetrics = Objects.requireNonNull(accountServiceMetrics, "accountServiceMetrics cannot be null");
+    this.accountConfig = Objects.requireNonNull(accountConfig, "accountConfig cannot be null");
+    this.scheduler = scheduler;
+
+    logger.info("Starting JSON file account service with file '{}'", accountFile.toString());
+  }
+
+  /** Guards against calling {@link #init} multiple times.*/
+  private final AtomicBoolean initialized = new AtomicBoolean();
+
+  /**
+   * Initializes this instance. Does all work which requires handing out references to this instance itself which can't
+   * be done during normal instance construction without letting a reference to an incomplete constructed instance escape.
+   */
+  void init() {
```

Comment:
In HelixAccountService we do similar initialization (with lambdas and passing references to the instance under construction) within the constructor. Is there an issue that arises when doing that?

Reply:
Ah, I found this post: https://stackoverflow.com/questions/2513597/what-is-an-incompletely-constructed-object

That is an interesting case. I guess it relates to the lambda's ability to see the final fields set in the constructor.

Reply:
Yes Lambda's added a new risk for letting the `this` reference "escape". It usually occurred when someone registered a callback from a constructor. The "no escaping" rule has been part of Java Concurrency in practice for a long time.